### PR TITLE
Implement Soft-Delete for Students in Classes to Improve Performance

### PR DESCRIPTION
**Pull request by phungmanhcuong**
To avoid long waiting from user when deleting student from class, need apply soft delete  
To enhance user experience and minimize wait times when removing a student from the class, implementing a soft delete feature is crucial.

### CODE REVIEW

File: db/migrate/20241014031756_add_deleted_at_to_course_users.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class AddDeletedAtToCourseUsers < ActiveRecord::Migration[7.0]
+  def change
+    add_column :course_users, :deleted_at, :datetime
```
Comment (cysjonathan) [inline db/migrate/20241014031756_add_deleted_at_to_course_users.rb:3]: Can be done in a single line: `add_column ... index: true`  
Separately, can explicity specify `default: nil`?

Also can put everything in a single migration file. Makes it easier for feature rollback if need be.

File: db/migrate/20241014031828_add_deleted_at_to_course_reference_timelines.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class AddDeletedAtToCourseReferenceTimelines < ActiveRecord::Migration[7.0]
+  def change
+    add_column :course_reference_timelines, :deleted_at, :datetime
```
Comment (cysjonathan) [inline db/migrate/20241014031828_add_deleted_at_to_course_reference_timelines.rb:3]: Why does Course Reference Timeline need soft-delete for course-user deletion?

File: db/migrate/20241014031821_add_deleted_at_to_course_groups.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class AddDeletedAtToCourseGroups < ActiveRecord::Migration[7.0]
+  def change
+    add_column :course_groups, :deleted_at, :datetime
```
Comment (cysjonathan) [inline db/migrate/20241014031821_add_deleted_at_to_course_groups.rb:3]: Why does Course Groups need soft-delete for course-user deletion?

File: db/migrate/20241014031810_add_deleted_at_to_course_achievements.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class AddDeletedAtToCourseAchievements < ActiveRecord::Migration[7.0]
+  def change
+    add_column :course_achievements, :deleted_at, :datetime
```
Comment (cysjonathan) [inline db/migrate/20241014031810_add_deleted_at_to_course_achievements.rb:3]: Why does Course Achievements need soft-delete for course-user deletion?

File: db/schema.rb [hunk 5]
```diff
@@ -1392,7 +1412,7 @@
     t.string "type", limit: 255, null: false
     t.string "name", limit: 255, null: false
     t.integer "parent_id"
-    t.serial "weight"
+    t.serial "weight", null: false
```
Comment (cysjonathan) [inline db/schema.rb:1415]: Don't change this line (remove the change) since this is not related to this PR.

File: app/models/course/group_user.rb [hunk 6]
```diff
@@ -1,8 +1,8 @@
 # frozen_string_literal: true
 class Course::GroupUser < ApplicationRecord
   after_initialize :set_defaults, if: :new_record?
-
-  enum :role, { normal: 0, manager: 1 }
+  acts_as_paranoid
+  enum role: { normal: 0, manager: 1 }
```
Comment (cysjonathan) [inline app/models/course/group_user.rb:5]: Why are you reverting a change from rails 7.2 upgrade PR?

https://github.com/Coursemology/coursemology2/blame/8e2a6b9cc8c402da6884c212f044ea04d407b314/app/models/course/group_user.rb#L5

File: spec/models/course_user_spec.rb [hunk 7]
```diff
@@ -164,6 +164,47 @@
       end
     end
 
+    describe 'soft delete behavior' do
```
Comment (cysjonathan) [inline spec/models/course_user_spec.rb:167]: Your tests does not cover checks that associated models are deleted and restored?

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -19,4 +21,29 @@ def other_instances
...
+        format.html { redirect_to course_users_path(@course), notice: 'Student was successfully removed from the course.' }
+        format.json { head :no_content }
+      end
+    else
+      respond_to do |format|
+        format.html { redirect_to course_users_path(@course), alert: 'Only students can be removed.' }
+        format.json { render json: { error: 'Only students can be removed' }, status: :unprocessable_entity }
+      end
+    end
+  end
```
Comment (cysjonathan) [inline app/controllers/users_controller.rb]: Deletion should apply for _any_ course_user, not just student.  
There is no business logic which says only students can be removed.

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -19,4 +21,29 @@ def other_instances
...
+
+  private
+
+  def set_course
+    @course = Course.find(params[:course_id])
+  end
+
+  def set_course_user
+    @course_user = @course.course_users.find(params[:id])
+  end
```
Comment (cysjonathan) [inline app/controllers/users_controller.rb]: Your logic should be modify `Course::UsersController`, not top-level `UsersController`.  
You should not be modifying User.

I just saw that you removed this change in the next commit. You should squash this changes as I review commit-by-commit.

File: Gemfile [hunk 10]
```diff
@@ -15,7 +15,7 @@ gem 'tzinfo-data', platforms: [:mswin, :mswin64]
 gem 'rails', '~> 7.2.1'
 
 # Soft delete
-gem 'paranoia', '~> 2.4'
```
Comment (cysjonathan) [inline Gemfile]: No point having two separate commits for  
`gem 'paranoia', '~> 2.4'`  
and  
`gem 'paranoia'`

Should squash into one.

**Review by cysjonathan — CHANGES_REQUESTED** [review]

File: app/models/course/group_user.rb (hunk 6 above)
Comment (phungmanhcuong) [reply app/models/course/group_user.rb:5]: Lemme revert it.

File: db/migrate/20241014031756_add_deleted_at_to_course_users.rb (hunk 1 above)
Comment (phungmanhcuong) [reply db/migrate/20241014031756_add_deleted_at_to_course_users.rb:3]: Actually, this is the generated file after running `rails db:migrate`. Can we skip it ?

File: db/migrate/20241014031810_add_deleted_at_to_course_achievements.rb (hunk 4 above)
Comment (phungmanhcuong) [reply db/migrate/20241014031810_add_deleted_at_to_course_achievements.rb:3]: Actually, I am not sure about the logic. So this is just proposal change.

File: db/schema.rb (hunk 5 above)
Comment (phungmanhcuong) [reply db/schema.rb:1415]: Ok

File: spec/models/course_user_spec.rb (hunk 7 above)
Comment (phungmanhcuong) [reply spec/models/course_user_spec.rb:167]: Ok, it should have

File: db/migrate/20241014031821_add_deleted_at_to_course_groups.rb (hunk 3 above)
Comment (phungmanhcuong) [reply db/migrate/20241014031821_add_deleted_at_to_course_groups.rb:3]: From this [line](https://github.com/Coursemology/coursemology2/blob/f84dfd0b3faface9645177ccafd93ca80e694187/app/models/course_user.rb#L58) I am assuming need delete it if delete course-user.

File: db/migrate/20241014031828_add_deleted_at_to_course_reference_timelines.rb (hunk 2 above)
Comment (phungmanhcuong) [reply db/migrate/20241014031828_add_deleted_at_to_course_reference_timelines.rb:3]: I will revert this one

File: db/migrate/20241014031821_add_deleted_at_to_course_groups.rb (hunk 3 above)
Comment (cysjonathan) [reply db/migrate/20241014031821_add_deleted_at_to_course_groups.rb]: That's Course::GroupUser not Course::Group right?

File: db/migrate/20241014031756_add_deleted_at_to_course_users.rb (hunk 1 above)
Comment (cysjonathan) [reply db/migrate/20241014031756_add_deleted_at_to_course_users.rb]: How does `rails db:migrate` generate the migration file?  
Did you mean `rails generate migration xxx`?

File: db/migrate/20241014031810_add_deleted_at_to_course_achievements.rb (hunk 4 above)
Comment (cysjonathan) [reply db/migrate/20241014031810_add_deleted_at_to_course_achievements.rb]: `Course::UserAchievement` is not the same as `Course::Achievement`

File: db/migrate/20241014031756_add_deleted_at_to_course_users.rb (hunk 1 above)
Comment (phungmanhcuong) [reply db/migrate/20241014031756_add_deleted_at_to_course_users.rb]: Yes, you are correct.

Comment (cysjonathan) [reply db/migrate/20241014031756_add_deleted_at_to_course_users.rb]: After generating the migration files, any time before executing the migration with `rails db:migrate`, you can edit the migration file yourself.

File: db/migrate/20241014031821_add_deleted_at_to_course_groups.rb (hunk 3 above)
Comment (phungmanhcuong) [reply db/migrate/20241014031821_add_deleted_at_to_course_groups.rb]: Yes, thats correct. I have reverted this change, its outdated now.

File: db/migrate/20241014031810_add_deleted_at_to_course_achievements.rb (hunk 4 above)
Comment (phungmanhcuong) [reply db/migrate/20241014031810_add_deleted_at_to_course_achievements.rb]: I reverted this one.

File: db/migrate/20241014031756_add_deleted_at_to_course_users.rb (hunk 1 above)
Comment (phungmanhcuong) [reply db/migrate/20241014031756_add_deleted_at_to_course_users.rb]: Ok, I updated for all migration files.

File: db/migrate/20241021110207_apply_soft_delete_for_course_user_and_associated_models.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
+class ApplySoftDeleteForCourseUserAndAssociatedModels1 < ActiveRecord::Migration[7.2]
```
Comment (cysjonathan) [inline db/migrate/20241021110207_apply_soft_delete_for_course_user_and_associated_models.rb:1]: `NameError: uninitialized constant ApplySoftDeleteForCourseUserAndAssociatedModels (NameError)`

File: tests/tests/users/sign-up.spec.ts [hunk 12]
```diff
@@ -101,6 +101,7 @@ test.describe('user invited to 2 courses', () => {
 
     await page.goto(`/courses/${course1.id}`);
     await expect(page).toHaveURL(new RegExp(course1.id));
+    await page.waitForTimeout(1000);
```
Comment (cysjonathan) [inline tests/tests/users/sign-up.spec.ts:104]: Does this solve a separate problem? This does not seem related to soft-deletion of course_user.

Fixing flaky tests should be in a separate PR, and you aren't explaining here why we now need to have this `waitForTimeout`.  
https://playwright.dev/docs/api/class-page#page-wait-for-timeout quoting the docs:  
> Tests using the timer in production are going to be flaky. Use signals such as network events, selectors becoming visible and others instead.

File: spec/controllers/course/admin/admin_controller_spec.rb [hunk 13]
```diff
@@ -111,9 +111,9 @@
       context 'when the user is a Course Manager' do
         let(:user) { create(:course_manager, course: course).user }
 
-        it 'destroys the course' do
+        it 'soft deletes the course' do
```
Comment (cysjonathan) [inline spec/controllers/course/admin/admin_controller_spec.rb:114]: Why do we need course soft-deletion in this PR?

File: db/migrate/20241021110207_apply_soft_delete_for_course_user_and_associated_models.rb [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+      t.datetime :deleted_at, default: nil unless column_exists?(:courses, :deleted_at)
+      t.index :deleted_at unless index_exists?(:courses, :deleted_at)
+    end
+
+    change_table :course_achievements do |t|
+      t.datetime :deleted_at, default: nil unless column_exists?(:course_achievements, :deleted_at)
+      t.index :deleted_at unless index_exists?(:course_achievements, :deleted_at)
+    end
+  end
+end
```
Comment (cysjonathan) [inline db/migrate/20241021110207_apply_soft_delete_for_course_user_and_associated_models.rb:48]: You can consider adding up and down migration steps for `db:migrate` and `db:rollback` respectively.

File: app/models/course_user.rb [hunk 15]
```diff
@@ -23,6 +24,15 @@ class CourseUser < ApplicationRecord
   # A set of roles which comprise the managers of a course.
   MANAGER_ROLES = Set[:manager, :owner].map { |v| roles[v] }.freeze
 
+  ASSOCIATED_MODELS = %i[
```
Comment (cysjonathan) [inline app/models/course_user.rb:27]: Are these associated models exhaustive? How about course_discussion_posts, submissions, video_submissions, activities, etc?

Are we sure we want to cascade delete / soft-delete the associated models in the first place?

File: app/models/course_user.rb [hunk 16]
```diff
@@ -176,6 +186,22 @@ def self.user?(user)
     all.exists?(user: user)
   end
 
+  # Soft delete the course user.
+  def destroy
```
Comment (cysjonathan) [inline app/models/course_user.rb:190]: Soft-delete of course-user with a lot of activities still take >5 minutes. Your PR does not resolve the root issue. Please revise.

File: app/models/course_user.rb (hunk 15 above)
Comment (cysjonathan) [reply app/models/course_user.rb:27]: Please confirm if you have also checked the effects on jobs or not?

File: Gemfile [hunk 17]
```diff
@@ -14,6 +14,9 @@ gem 'tzinfo-data', platforms: [:mswin, :mswin64]
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
 gem 'rails', '~> 7.2.1'
 
+# Soft delete
+gem 'paranoia'
```
Comment (cysjonathan) [inline Gemfile:18]: The `paranoia` gem github page also recommended to look at `discard` gem

> paranoia has some surprising behaviour (like overriding ActiveRecord's delete and destroy) and is not recommended for new projects. See [discard's README](https://github.com/jhawthorn/discard#why-not-paranoia-or-acts_as_paranoid) for more details.

Have you taken a look at `discard`?

File: app/models/course_user.rb [hunk 18]
```diff
@@ -176,6 +186,22 @@ def self.user?(user)
...
+    ASSOCIATED_MODELS.each do |association|
+      send(association).destroy_all
+    end
+    super
+  end
+
+  # Restore the course user.
+  def restore
```
Comment (cysjonathan) [inline app/models/course_user.rb:198]: When I try re-inviting user into a course, I get:
`PG::UniqueViolation: ERROR: duplicate key value violates unique constraint "index_course_users_on_course_id_and_user_id"
DETAIL: Key (course_id, user_id)=(course-xxx, user-yyy) already exists.`

Have you tested that users can be restored from the webapp?  

- instructor invites user into course using the Invite Users page
- user self-enrolls using course registration code (requires approval) from the Course Home page

**Review by cysjonathan — CHANGES_REQUESTED** [review]

File: tests/tests/users/sign-up.spec.ts (hunk 12 above)
Comment (phungmanhcuong) [reply tests/tests/users/sign-up.spec.ts:104]: Ok, its not related. Lemme revert it.

File: db/migrate/20241021110207_apply_soft_delete_for_course_user_and_associated_models.rb (hunk 14 above)
Comment (phungmanhcuong) [reply db/migrate/20241021110207_apply_soft_delete_for_course_user_and_associated_models.rb:48]: Ok

File: spec/controllers/course/admin/admin_controller_spec.rb (hunk 13 above)
Comment (phungmanhcuong) [reply spec/controllers/course/admin/admin_controller_spec.rb:114]: Because we apply the soft-delete for CourseUser, which depend on Course.
We also allow to delete the Course, but if the Course is deleted, the CourseUser will need hard-delete to avoid "FK" violation.
So we need soft-delete for Course also.

File: app/models/course_user.rb (hunk 18 above)
Comment (phungmanhcuong) [reply app/models/course_user.rb:198]: I have tested the scenario ``` Remove student from course and then enrol him again```.
`re-inviting user` The `invite` API has been updated.

File: app/models/course_user.rb (hunk 16 above)
Comment (phungmanhcuong) [reply app/models/course_user.rb:190]: The only way is verify by similar data volumn on production. We can keep monitoring it on staging to see.
In theory, the soft-delete won't effect to the index and fk-constrains, so it would be much faster than hard-delete.

Comment (cysjonathan) [reply app/models/course_user.rb:190]: I already verified the slow deletion using production-grade data, the problem is that there are many individual associated updates deletes involved.  

I suggest we simply do soft-delete on course-user and do not touch associated models.  
Subsequent calls to query data should ensure scope of course-user does not include soft-deleted course-user. `discard` gem might be better for this.

File: app/models/course_user.rb (hunk 15 above)
Comment (phungmanhcuong) [reply app/models/course_user.rb:27]: I have updated to skip cascading soft-delete on associated models, so we don't need to worry about side effects from this.

File: app/models/course_user.rb (hunk 18 above)
Comment (phungmanhcuong) [reply app/models/course_user.rb]: The API for "invitation" and "registration by code" has been updated. So all the scenarios to add user to course should be works now.

File: db/migrate/20241021110207_apply_soft_delete_for_course_user_and_associated_models.rb (hunk 11 above)
Comment (phungmanhcuong) [reply db/migrate/20241021110207_apply_soft_delete_for_course_user_and_associated_models.rb]: You get this error when run "rails db:migrate" right ?

File: Gemfile (hunk 17 above)
Comment (phungmanhcuong) [reply Gemfile:18]: Learning the discard.

File: spec/controllers/course/admin/admin_controller_spec.rb (hunk 13 above)
Comment (cysjonathan) [reply spec/controllers/course/admin/admin_controller_spec.rb:114]: We can hard-delete CourseUser if associated Course is deleted.  
However, if we want to implement soft-delete for Course, we should also ensure that the soft-deletion is performant + there must be a way to restore the course. We can do this in a separate PR.

File: app/models/course_user.rb (hunk 16 above)
Comment (phungmanhcuong) [reply app/models/course_user.rb]: Spend time researching about discard. Might apply it later then.

File: spec/controllers/course/admin/admin_controller_spec.rb (hunk 13 above)
Comment (phungmanhcuong) [reply spec/controllers/course/admin/admin_controller_spec.rb:114]: The soft-delete for Course PR might need come first this PR, otherwise, a few of tests (which delete the Course) will be failed.

File: app/models/course_user.rb (hunk 15 above)
Comment (cysjonathan) [reply app/models/course_user.rb:27]: This variable `ASSOCIATED_MODELS` is unused, it should not exist.

File: app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb [hunk 19]
```diff
@@ -79,15 +79,17 @@ def restrict_invitee_role(users)
...
+
       name = value[:name].presence || value[:email]
       phantom = ActiveRecord::Type::Boolean.new.cast(value[:phantom])
       { name: name,
         email: value[:email],
         role: value[:role],
         phantom: phantom,
         timeline_algorithm: value[:timeline_algorithm] }
-    end
+    end.compact
```
Comment (cysjonathan) [inline app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb:92]: Why is there this change here?

File: db/migrate/20241022035102_apply_soft_delete_for_course_user.rb [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+    end
+
+    change_table :courses do |t|
+      t.datetime :deleted_at, default: nil unless column_exists?(:courses, :deleted_at)
+      t.index :deleted_at unless index_exists?(:courses, :deleted_at)
+    end
+
+    change_table :course_achievements do |t|
+      t.datetime :deleted_at, default: nil unless column_exists?(:course_achievements, :deleted_at)
+      t.index :deleted_at unless index_exists?(:course_achievements, :deleted_at)
```
Comment (cysjonathan) [inline db/migrate/20241022035102_apply_soft_delete_for_course_user.rb:15]: Why is there still `deleted_at` on `course_achievements` here?

File: db/schema.rb [hunk 21]
```diff
@@ -1392,7 +1398,7 @@
...
```
Comment (cysjonathan) [inline db/schema.rb:1401]: This issue again: Don't change this line (remove the change) since this is not related to this PR.

File: spec/controllers/course/admin/admin_controller_spec.rb (hunk 13 above)
Comment (cysjonathan) [reply spec/controllers/course/admin/admin_controller_spec.rb:114]: > We can hard-delete CourseUser if associated Course is deleted.

**Review by cysjonathan — CHANGES_REQUESTED** [review]
You still need to squash your commits before asking for review please.
